app/services/change_offer.rb

```diff
@@ -12,21 +14,37 @@ def initialize(actor:,
   end
 
   def save!
-    change_an_offer = ChangeAnOffer.new(actor: actor,
-                                        application_choice: application_choice,
-                                        course_option: course_option,
-                                        offer_conditions: conditions)
-    unless change_an_offer.save
-      if change_an_offer.errors[:base].include?('The new offer is identical to the current offer')
-        raise IdenticalOfferError, 'The new offer is identical to the current offer'
-      elsif change_an_offer.errors[:course_option].present?
-        raise CourseValidationError, change_an_offer.errors[:course_option].join
-      else
-        raise 'Unable to complete save on change_an_offer'
+    auth.assert_can_make_decisions!(application_choice: application_choice, course_option: course_option)
+
+    if offer.valid?
+      audit(actor) do
+        ActiveRecord::Base.transaction do
+          ApplicationStateChange.new(application_choice).make_offer!
+
+          application_choice.current_course_option = course_option
+          application_choice.offer = { 'conditions' => conditions }
+          application_choice.offer_changed_at = Time.zone.now
+          application_choice.save!
+        end
+
+        SetDeclineByDefault.new(application_form: application_choice.application_form).call
```

Comment:
I think this should probably be wrapped within the transaction as well.

Reply:
Yea as long as this doesn't trigger any jobs

Reply:
It only updates the decline_by_default date and number of dates, so it makes sense to rollback the entire transaction if it fails.